Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp

```diff
@@ -0,0 +1,386 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "AudioEncoderCocoa.h"
+
+#if ENABLE(WEB_CODECS) && USE(AVFOUNDATION)
+
+#include "AudioDecoderCocoa.h"
+#include "AudioSampleBufferConverter.h"
+#include "AudioSampleFormat.h"
+#include "CAAudioStreamDescription.h"
+#include "MediaSampleAVFObjC.h"
+#include "MediaUtilities.h"
+#include "PlatformRawAudioDataCocoa.h"
+#include "WebMAudioUtilitiesCocoa.h"
+
+#include <CoreAudio/CoreAudioTypes.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/TZoneMallocInlines.h>
+#include <wtf/ThreadSafeRefCounted.h>
+#include <wtf/WorkQueue.h>
+#include <wtf/text/MakeString.h>
+
+#include <pal/cf/CoreMediaSoftLink.h>
+
+namespace WebCore {
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(AudioEncoderCocoa);
+
+class InternalAudioEncoderCocoa : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<InternalAudioEncoderCocoa, WTF::DestructionThread::Main> {
+    WTF_MAKE_TZONE_ALLOCATED_INLINE(InternalAudioEncoderCocoa);
+    WTF_MAKE_NONCOPYABLE(InternalAudioEncoderCocoa);
+
+public:
+    using Config = AudioEncoder::Config;
+    using EncodePromise = AudioEncoder::EncodePromise;
+
+    static Ref<InternalAudioEncoderCocoa> create(const Config& config, AudioEncoder::DescriptionCallback&& descriptionCallback, AudioEncoder::OutputCallback&& outputCallback)
+    {
+        return adoptRef(*new InternalAudioEncoderCocoa(config, WTFMove(descriptionCallback), WTFMove(outputCallback)));
+    }
+    ~InternalAudioEncoderCocoa() = default;
+
+    String initialize(const String& codecName);
+    Ref<AudioEncoder::EncodePromise> encode(AudioEncoder::RawFrame&&);
+    Ref<GenericPromise> flush();
+    void reset();
+    void close();
+
+    static WorkQueue& queueSingleton() { return AudioDecoderCocoa::queueSingleton(); }
+
+private:
+    InternalAudioEncoderCocoa(const AudioEncoder::Config&, AudioEncoder::DescriptionCallback&&, AudioEncoder::OutputCallback&&);
+    static void compressedAudioOutputBufferCallback(void*, CMBufferQueueTriggerToken);
+    Ref<AudioSampleBufferConverter> converter() const { return *m_converter; }
+    void processEncodedOutputs();
+    AudioEncoder::ActiveConfiguration activeConfiguration(CMSampleBufferRef) const;
+    Vector<uint8_t> generateDecoderDescriptionFromSample(CMSampleBufferRef) const;
+
+    AudioEncoder::DescriptionCallback m_descriptionCallback;
+    AudioEncoder::OutputCallback m_outputCallback;
+    RefPtr<AudioSampleBufferConverter> m_converter;
+
+    bool m_isClosed WTF_GUARDED_BY_CAPABILITY(queueSingleton()) { false };
+    const AudioEncoder::Config m_config;
+    String m_codecString; // set during initialisation, const after.
```

Comment:
It is not great to use strings with work queues, I would tend to use an enum instead.

Reply:
I had planned to do it in a followup

Reply:
I've removed it fully, WebCodecsAudioEncoder turned out to copy the original string if none is provided in the new description